provider/parameter.go

```diff
@@ -429,84 +416,170 @@ func (v *Parameter) Valid() diag.Diagnostics {
-			if exists {
-				return diag.Diagnostics{
-					{
-						Severity: diag.Error,
-						Summary:  "Option values must be unique.",
-						Detail:   fmt.Sprintf("multiple options found with the same value %q", option.Value),
-					},
-				}
-			}
-			diags := valueIsType(optionType, option.Value, cty.Path{})
-			if diags.HasError() {
-				return diags
-			}
-			optionValues[option.Value] = nil
-			optionNames[option.Name] = nil
+	optionValues, diags := v.ValidOptions(optionType)
+	if diags.HasError() {
+		return "", diags
+	}
+
+	// TODO: This is a bit of a hack. The current behavior states if validation
+	//   is given, then apply validation to unset values.
+	//   value == nil should not be accepted in the first place.
+	//   To fix this, value should be coerced to an empty string
+	//   if it is nil. Then let the validation logic always apply.
+	if len(v.Validation) == 0 && value == nil {
+		return "", nil
+	}
+
+	// forcedValue ensures the value is not-nil.
+	var forcedValue string
+	if value != nil {
+		forcedValue = *value
+	}
+
+	d := v.validValue(forcedValue, optionType, optionValues, cty.Path{})
+	if d.HasError() {
+		return "", d
+	}
+
+	err = valueIsType(v.Type, forcedValue)
+	if err != nil {
+		return "", diag.Diagnostics{
+			{
+				Severity: diag.Error,
+				Summary:  fmt.Sprintf("Parameter value is not of type %q", v.Type),
+				Detail:   err.Error(),
+			},
 		}
 	}
 
-	if v.Default != "" && len(v.Option) > 0 {
+	return forcedValue, nil
+}
+
+func (v *Parameter) ValidOptions(optionType OptionType) (map[string]struct{}, diag.Diagnostics) {
+	optionNames := map[string]struct{}{}
+	optionValues := map[string]struct{}{}
+
+	var diags diag.Diagnostics
+	for _, option := range v.Option {
+		_, exists := optionNames[option.Name]
+		if exists {
+			return nil, diag.Diagnostics{{
+				Severity: diag.Error,
+				Summary:  "Option names must be unique.",
+				Detail:   fmt.Sprintf("multiple options found with the same name %q", option.Name),
+			}}
+		}
+
+		_, exists = optionValues[option.Value]
+		if exists {
+			return nil, diag.Diagnostics{{
+				Severity: diag.Error,
+				Summary:  "Option values must be unique.",
+				Detail:   fmt.Sprintf("multiple options found with the same value %q", option.Value),
+			}}
+		}
+
+		err := valueIsType(optionType, option.Value)
+		if err != nil {
+			diags = append(diags, diag.Diagnostic{
+				Severity: diag.Error,
+				Summary:  fmt.Sprintf("Option %q with value=%q is not of type %q", option.Name, option.Value, optionType),
+				Detail:   err.Error(),
+			})
+			continue
+		}
+		optionValues[option.Value] = struct{}{}
+		optionNames[option.Name] = struct{}{}
+
+		// Option values are assumed to be valid. Do not call validValue on them.
+	}
+
+	if diags != nil && diags.HasError() {
+		return nil, diags
+	}
+	return optionValues, nil
+}
+
+func (v *Parameter) validValue(value string, optionType OptionType, optionValues map[string]struct{}, path cty.Path) diag.Diagnostics {
```

Comment:
this function is called only once, with an empty path.

Reply:
Updated the call to pass in the default value path if the value is sourced from the default.

The path (if it can be specified), makes the error a bit nicer:

![Screenshot From 2025-05-07 07-50-11](https://github.com/user-attachments/assets/638cb44f-903c-4b0e-9f67-6edbd607aa0b)


Without path

![Screenshot From 2025-05-07 07-50-16](https://github.com/user-attachments/assets/15bce8a7-88f8-44ab-a3f8-445e5149c6c7)